algebra/vector_space.lean

```diff
@@ -12,197 +12,209 @@ lemma eq_zero_of_add_self_eq {α : Type} [add_group α]
-@[simp] lemma map_mul_simp : A.T (c•v) = c • (A.T v) := map_mul A c v
+instance : has_coe_to_fun (linear_space K V W) := ⟨_, T⟩
 
-theorem ext (HAB : ∀ v, A.T v = B.T v) : A = B :=
+@[simp] lemma map_add_vec : A (u+v) = A u + A v := map_add A u v
+@[simp] lemma map_smul_vec : A (c•v) = c • (A v) := map_smul A c v
+
+theorem ext (HAB : ∀ v, A v = B v) : A = B :=
 by {cases A, cases B, congr, exact funext HAB}
 
-@[simp] lemma map_zero : A.T 0 = 0 :=
+@[simp] lemma map_zero : A 0 = 0 :=
 begin
-    have := eq.symm (map_add A 0 0),
-    rw [add_zero] at this,
-    exact eq_zero_of_add_self_eq this
+  have := eq.symm (map_add A 0 0),
+  rw [add_zero] at this,
+  exact eq_zero_of_add_self_eq this
 end
 
-@[simp] lemma map_neg : A.T (-v) = -(A.T v) :=
-eq_neg_of_add_eq_zero (by {rw [←map_add], simp})
+@[simp] lemma map_neg : A (-v) = -(A v) :=
+eq_neg_of_add_eq_zero (by {rw [←map_add_vec], simp})
 
-@[simp] lemma map_sub : A.T (u-v) = A.T u - A.T v := by simp
+@[simp] lemma map_sub : A (u-v) = A u - A v := by simp
 
 end basic
 
 
-@[simp] def ker (A : linear_space K V W) : V → Prop := (λ v, A.T v = 0)
+@[simp] def ker (A : linear_space K V W) : set V := {v | A v = 0}
 
 namespace ker
 
 variables {A : linear_space K V W}
-variables (u v : V)
+variables {u v : V}
+
+@[simp] lemma vec_ker : v ∈ A.ker ↔ A v = 0 := ⟨id, id⟩
 
-theorem ker_of_map_eq_map (Huv : A.T u = A.T v) : A.ker (u-v) :=
-by {rw [ker,map_sub], exact sub_eq_zero_of_eq Huv}
+theorem ker_of_map_eq_map (Huv : A u = A v) : u-v ∈ A.ker :=
+by {rw [vec_ker,map_sub], exact sub_eq_zero_of_eq Huv}
 
-theorem inj_of_trivial_ker (H: ∀ v, A.ker v → v = 0) (Huv: A.T u = A.T v) : u = v :=
-eq_of_sub_eq_zero (H _ (ker_of_map_eq_map _ _ Huv))
+theorem inj_of_trivial_ker (H: ∀ v, A.ker v → v = 0) (Huv: A u = A v) : u = v :=
+eq_of_sub_eq_zero (H _ (ker_of_map_eq_map Huv))
 
-theorem ker_add (HU : A.ker u) (HV : A.ker v) : A.ker (u + v) :=
-by {unfold ker at *, simp [HU,HV]}
+variables (u v)
 
-theorem ker_zero : A.ker 0 := map_zero
+theorem add_ker (HU : u ∈ A.ker) (HV : v ∈ A.ker) : u + v ∈ A.ker :=
+by {rw [vec_ker] at *, simp [HU,HV]}
 
-theorem ker_neg (HV : A.ker v) : A.ker (-v) :=
-by {unfold ker at *, simp [HV]}
+theorem zero_ker : (0:V) ∈ A.ker := map_zero
 
-theorem ker_sub (HU : A.ker u) (HV : A.ker v) : A.ker (u - v) :=
-by {unfold ker at *, simp [HU,HV]}
+theorem neg_ker (HV : v ∈ A.ker) : -v ∈ A.ker :=
+by {rw [vec_ker] at *, simp [HV]}
 
-theorem ker_smul (c : K) (v : V) (HV : A.ker v) : A.ker (c • v) :=
-by {unfold ker at *, simp [HV]}
+theorem sub_ker (HU : u ∈ A.ker) (HV : v ∈ A.ker) : u - v ∈ A.ker :=
+by {rw [vec_ker] at *, simp [HU,HV]}
+
+theorem smul_ker (c : K) (v : V) (HV : v ∈ A.ker) : c • v ∈ A.ker :=
+by {rw [vec_ker] at *, simp [HV]}
 
 instance : subspace K V A.ker :=
-{
-    p_add := ker_add,
-    p_zero := ker_zero,
-    p_neg := ker_neg,
-    p_smul := ker_smul
-}
+{ add  := add_ker,
+  zero := zero_ker,
+  neg  := neg_ker,
+  smul := smul_ker }
```

Comment:
I suggest you inline these proofs, so that people can just use the subspace names when applying `zero_ker` etc.